src/main/java/com/mongodb/spark/sql/connector/read/MongoMicroBatchPartitionReader.java

```diff
@@ -0,0 +1,162 @@
+
+/**
+ * A partition reader returned by {@link
+ * MongoPartitionReaderFactory#createReader(org.apache.spark.sql.connector.read.InputPartition)}.
+ * It's responsible for outputting data for a RDD partition.
+ */
+public class MongoMicroBatchPartitionReader implements PartitionReader<InternalRow> {
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(MongoMicroBatchPartitionReader.class);
+  private static final String FULL_DOCUMENT = "fullDocument";
+  private final MongoMicroBatchInputPartition partition;
+  private final BsonDocumentToRowConverter bsonDocumentToRowConverter;
+  private final ReadConfig readConfig;
+  private boolean closed = false;
+  private MongoClient mongoClient;
+  private MongoChangeStreamCursor<BsonDocument> changeStreamCursor;
+  private InternalRow currentRow;
+
+  /**
+   * Construct a new instance
+   *
+   * @param partition the partition
+   * @param bsonDocumentToRowConverter the converter from {@link BsonDocument} to {@link
+   *     InternalRow}
+   * @param readConfig the read configuration for reading from the partition
+   */
+  public MongoMicroBatchPartitionReader(
+      final MongoMicroBatchInputPartition partition,
+      final BsonDocumentToRowConverter bsonDocumentToRowConverter,
+      final ReadConfig readConfig) {
+
+    this.partition = partition;
+    this.bsonDocumentToRowConverter = bsonDocumentToRowConverter;
+    this.readConfig = readConfig;
+    LOGGER.info(
+        "Creating partition reader for: PartitionId: {} with Schema: {}",
+        partition.getPartitionId(),
+        bsonDocumentToRowConverter.getSchema());
+  }
+
+  /** Proceed to next record, returns false if there is no more records. */
+  @Override
+  public boolean next() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call next() on a closed PartitionReader.");
+    MongoChangeStreamCursor<BsonDocument> cursor = getCursor();
+    BsonDocument next = cursor.tryNext();
+    if (next != null) {
+      if (readConfig.streamPublishFullDocumentOnly()) {
+        next = next.getDocument(FULL_DOCUMENT, new BsonDocument());
+      }
+      currentRow = bsonDocumentToRowConverter.toInternalRow(next);
+    }
+    return next != null;
+  }
+
+  /** Return the current record. This method should return same value until `next` is called. */
+  @Override
+  public InternalRow get() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call get() on a closed PartitionReader.");
+    return currentRow;
+  }
+
+  @Override
+  public void close() {
+    if (!closed) {
+      closed = true;
+      releaseCursor();
+    }
+  }
+
+  private MongoChangeStreamCursor<BsonDocument> getCursor() {
+    if (mongoClient == null) {
+      mongoClient = readConfig.getMongoClient();
+    }
+    if (changeStreamCursor == null) {
+
+      List<BsonDocument> pipeline = new ArrayList<>();
+      pipeline.add(
+          Aggregates.match(Filters.lt("clusterTime", partition.getEndOffsetTimestamp()))
```

Comment:
Confirm that 

- clusterTime corresponds to operationTime
- no events will be missed or duplicated

Reply:
Just added this as a note to self, as I don't fully understand that this is correct.

Reply:
Checked in #change-streams
> the clusterTime field of an event is exactly the same as the timestamp used in that event’s resume token